File: GrokkingDP/5/LongestRepeatingSubseq.cpp
#include <iostream>
#include <vector>

enum class Method { BRUTE, MEM, BU };

int LongestRepeatingSubseq(std::string str, Method m);
int Brute(std::string str, size_t i, size_t j);
int Memoize(std::string str, size_t i, size_t j,
            std::vector<std::vector<int>>& dp);
int BU(std::string str);

int main(void) {
  Method m = Method::MEM;
  std::cout << LongestRepeatingSubseq("tomorrow", m) << ":"
            << LongestRepeatingSubseq("aabdbcec", m) << ":"
            << LongestRepeatingSubseq("fmff", m) << ":"
            << LongestRepeatingSubseq("fmffnaijsdfnwer", m);
  return 0;
}

int LongestRepeatingSubseq(std::string str, Method m) {
  std::vector<std::vector<int>> dp(str.length(),
                                   std::vector<int>(str.length()));
  switch (m) {
    case Method::BRUTE:
      return Brute(str, 1, 0);
    case Method::MEM:
      return Memoize(str, 1, 0, dp);
    case Method::BU:
      return BU(str);
  }
}

int Brute(std::string str, size_t i, size_t j) {
  if (j >= i || i == str.length()) return 0;
  if (str[i] == str[j]) return 1 + Brute(str, i + 1, j + 1);
  return std::max(Brute(str, i + 1, j), Brute(str, i, j + 1));
}

int Memoize(std::string str, size_t i, size_t j,
            std::vector<std::vector<int>>& dp) {
  if (j >= i || i == str.length()) return 0;
  if (dp[i][j] == 0) {
    if (str[i] == str[j])
      dp[i][j] = 1 + Brute(str, i + 1, j + 1);
    else
      dp[i][j] = std::max(Brute(str, i + 1, j), Brute(str, i, j + 1));
  }
  return dp[i][j];
}

int BU(std::string str) {
  int len = str.length();
  std::vector<std::vector<int>> dp(len + 1,
                                   std::vector<int>(len + 1));
  int longest = 0;
  for (int i = 1; i <= len; ++i) {
    for (int j = i + 1; j <= len; ++j) {
      if (str[i - 1] == str[j - 1])
        dp[i][j] = dp[i - 1][j - 1] + 1;
      else
        dp[i][j] = std::max(dp[i - 1][j], dp[i][j - 1]);
    }
    longest = std::max(longest, dp[i][len]);
  }
  return longest;
}

File: GrokkingDP/5/longest_repeating_subseq.py
#  Problem Statement #
#  Given a sequence, find the length of its longest repeating subsequence (LRS). A repeating subsequence will be the one that appears at least twice in the original sequence and is not overlapping (i.e. none of the corresponding characters in the repeating subsequences have the same index).

#  Example 1:

#  Input: “t o m o r r o w”
#  Output: 2
#  Explanation: The longest repeating subsequence is “or” {tomorrow}.

#  Example 2:

#  Input: “a a b d b c e c”
#  Output: 3
#  Explanation: The longest repeating subsequence is “a b c” {a a b d b c e c}.

#  Example 3:

#  Input: “f m f f”
#  Output: 2
#  Explanation: The longest repeating subsequence is “f f” {f m f f, f m f f}. Please note the second last character is shared in LRS.
s1, s2, s3 = 'tomorrow', 'aabdbcec', 'fmff'

def td(s: str) -> int:
    """
    >>> td(s1)
    2
    >>> td(s2)
    3
    >>> td(s3)
    2
    """
    S = len(s)
    dp = [[None for _ in range(S)] for _ in range(S)]
    def fn(i: int, j: int) -> int:
        if i == S or j == S: return 0
        if dp[i][j] == None:
            if i != j and s[i] == s[j]: dp[i][j] = 1 + fn(i+1, j+1)
            else: dp[i][j] = max(fn(i+1, j), fn(i, j+1))
        return dp[i][j]
    return fn(0, 0)

def bu(s: str) -> int:
    """
    >>> bu(s1)
    2
    >>> bu(s2)
    3
    >>> bu(s3)
    2
    """
    S = len(s)
    dp = [0] * S
    for i in range(1, S):
        for j in range(i):
            dp[i] = max(dp[i], dp[j])
            if s[i] == s[j]: dp[i] += 1
    return dp[-1]

if __name__ == '__main__':
    import doctest
    doctest.testmod()
